pep-0705.rst

```diff
@@ -0,0 +1,279 @@
+
+Discussion:
+
+* Value types behave covariantly, since ``TypedMapping`` objects have no mutator methods. This is similar to container types such as ``Mapping``, and different from relationships between two ``TypedDict`` types. Example::
+
+    class A(TypedMapping):
+        x: int | None
+
+    class B(TypedDict):
+        x: int
+
+    def f(a: A) -> None:
+        print(a['x'] or 0)
+
+    b: B = {'x': 0}
+    f(b)  # Accepted by type checker
+
+* A ``TypedDict`` or ``TypedMapping`` type with a required key is consistent with a ``TypedMapping`` type where the same key is a non-required key, again unlike relationships between two ``TypedDict`` types. Example::
+
+    class A(TypedMapping, total=False):
+        x: int
+
+    class B(TypedDict):
+        x: int
+
+    def f(a: A) -> None:
+        print(a.get('x', 0))
+
+    b: B = {'x': 0}
+    f(b)  # Accepted by type checker
+
+* A ``TypedMapping`` type ``A`` with no key ``'x'`` is not consistent with a ``TypedMapping`` type with a non-required key ``'x'``, since at runtime the key ``'x'`` could be present and have an incompatible type (which may not be visible through ``A`` due to structural subtyping). This is the same as for ``TypedDict`` types. Example::
+
+    class A(TypedMapping, total=False):
+        x: int
+        y: int
+
+    class B(TypedMapping, total=False):
+        x: int
+
+    class C(TypedMapping, total=False):
+        x: int
+        y: str
+
+    def f(a: A) -> None:
+        print(a.get('y') + 1)
+
+    def g(b: B) -> None:
+        f(b)  # Type check error: 'B' incompatible with 'A'
+
+    c: C = {'x': 0, 'y': 'foo'}
+    g(c)  # Runtime error: str + int
+
+* A ``TypedMapping`` with all ``int`` values is not consistent with ``Mapping[str, int]``, since there may be additional non-``int`` values not visible through the type, due to structural subtyping. This mirrors ``TypedDict``. Example::
+
+    class A(TypedMapping):
+        x: int
+
+    class B(TypedMapping):
+        x: int
+        y: str
+
+    def sum_values(m: Mapping[str, int]) -> int:
+        return sum(m.values())
+
+    def f(a: A) -> None:
+        sum_values(a)  # Type check error: 'A' incompatible with Mapping[str, int]
+
+    b: B = {'x': 0, 'y': 'foo'}
+    f(b)  # Runtime error: int + str
+
+
+Backwards Compatibility
+=======================
+
+This PEP changes the rules for how ``TypedDict`` behaves (allowing subclasses to
+inherit from ``TypedMapping`` protocols in a way that changes the resulting
+overloads), so code that inspects ``TypedDict`` types will have to change. This
+is expected to mainly affect type-checkers.
+
+The ``TypedMapping`` type will be added to the ``typing_extensions`` module,
+enabling its use in older versions of Python. Users wanting to inherit from
+both ``TypedDict`` and ``TypedMapping`` will need to use
+``typing_extensions.TypedDict`` in these versions, as the ``typing`` declaration
+of ``TypedDict`` does an ``is`` check on the metaclass of all base classes.
```

Comment:
I would at least briefly explicitly mention why this is a problem.

Reply:
```suggestion
of ``TypedDict`` does an ``is`` check on the metaclass of all base classes and raises
an error if any other metaclass is found. This would cause any subclass of both
this implementation and a ``TypedMapping`` protocol to trigger an error at runtime.
```
How about this?

Reply:
Looks great, thanks! :+1: